File: magic.py
def createbox():
    global box
    box=[[0 for _ in range(4)] for _ in range(4)]
    box[0][0]=a0
    box[0][1]=b0
    box[0][2]=c0
    box[0][3]=d0
    return box
def createglobals():
    global a0,b0,c0,d0,x
    y=input('Enter Date of Birth(dd/mm/yyyy)  ::   ')
    while len(y)!=10:
        y=input('Please Enter Valid Date of Birth in (DD/MM/YYYY) pattern Ex. 01/01/2000  ::  \nPress 1 to exit  ::  ')
        if y=='1':
            return False
    a=y[0]+y[1]
    b=y[3]+y[4]
    c=y[6]+y[7]+y[8]+y[9]
    try:
        if a[0]=='0':
            a0=int(a[1])
        else:
            a0=int(a)
        if b[0]=='0':
            b0=int(b[1])
        else:
            b0=int(b)
        c0=int(c[0]+c[1])
        d0=int(c[2]+c[3])
        x=a0+b0+c0+d0
    except ValueError:
        print("You didn't enter Correct Date of Birth or Something Went Wrong\n\t\t PLEASE TRY AGAIN")
        return False
    return a0,b0,c0,d0,x
def printbox():
    for i in range(4):
        for j in range(4):
            print(box[i][j],end='   ')
        print('\n')
def checkDuplicate(a,n):
    for i in range(n):
        for j in range(4):
            if box[i][j]==a:
                return True
    return False
def checkSemiMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x):
    if a0+b0+c0+d0==x and a1+b1+c1+d1==x and a2+b2+c2+d2==x and a3+b3+c3+d3==x and a0+a1+a2+a3==x and b0+b1+b2+b3==x and c0+c1+c2+c3==x and d0+d1+d2+d3==x:
        return True
    return False
def checkOrdinaryMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x):
    if a0+b1+c2+d3==x and d0+c1+b2+a3==x:
        return True
    return False
def checkMPMS(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x):
    if a0+b0+a1+b1==x and b0+c0+b1+c1==x and c1+c0+d0+d1==x and a2+b2+a1+b1==x and c1+c2+b1+b2==x and c1+c2+d1+d2==x and a2+b2+a3+b3==x and c2+c3+d2+d3==x and b2+b3+c2+c3==x and a0+b0+a3+b3==x and b0+c0+b3+c3==x and c0+d0+c3+d3==x and a0+a1+d0+d1==x and a1+a2+d1+d2==x and a2+a3+d2+d3==x:
        return True
    return False
def checkPanMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x):
    if c0+b1+a2+d3==x and d1+c2+b3+a0==x and b0+c1+d2+a3==x and a1+b2+c3+d0==x:
        return True
    return False
def checkCompleteMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x):
    if a0+c0+a2+c2==x and b0+d0+b2+d2==x and a1+c1+a3+c3==x and b1+d1+b3+d3==x:
        return True
    return False
def createMagic():
    global l
    z=False;l=[]
    while box[1][0]<x-(box[0][0]+box[0][1]):
        box[1][1]=x-(box[0][0]+box[0][1]+box[1][0])
        box[1][2]=x-(box[0][2]+box[1][1]+box[0][1])
        box[1][3]=x-(box[0][3]+box[1][2]+box[0][2])
        l.extend([box[1][0],box[1][1],box[1][2],box[1][3]])
        if (checkDuplicate(l[0],1) or checkDuplicate(l[1],1) or checkDuplicate(l[2],1) or checkDuplicate(l[3],1)) or any(i<0 for i in l) or len(l)>len(set(l)):
            l.clear()
        else:   
            l.clear()
            box[2][0]=0
            while box[2][0]<x-(box[1][0]+box[1][1]):
                box[2][1]=x-(box[1][0]+box[1][1]+box[2][0])
                box[2][2]=x-(box[1][2]+box[2][1]+box[1][1])
                box[2][3]=x-(box[1][3]+box[2][2]+box[1][2])
                l.extend([box[2][0],box[2][1],box[2][2],box[2][3]])
                if (checkDuplicate(l[0],2) or checkDuplicate(l[1],2) or checkDuplicate(l[2],2) or checkDuplicate(l[3],2)) or any(i<0 for i in l) or len(l)>len(set(l)):
                    l.clear()
                else:
                    l.clear()
                    box[3][0]=x-(box[2][0]+box[1][0]+box[0][0])
                    box[3][1]=x-(box[2][1]+box[1][1]+box[0][1])
                    box[3][2]=x-(box[2][2]+box[1][2]+box[0][2])
                    box[3][3]=x-(box[2][3]+box[1][3]+box[0][3])
                    l.extend([box[3][0],box[3][1],box[3][2],box[3][3]])
                    if checkSemiMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x) and not (checkDuplicate(box[3][0],3) or checkDuplicate(box[3][1],3) or checkDuplicate(box[3][2],3) or checkDuplicate(box[3][3],3)) and not any(i<0 for i in l) and len(l)==len(set(l)):
                        z=True
                        break
                    else:
                        pass
                box[2][0]+=1
        if z==True:
            break
        else:
            box[1][0]+=1
def checkRemaining():
    checkSemiM=False;checkOrdinaryM=False;checkM=False;checkPanM=False;checkCompleteM=False
    if checkSemiMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x):
        print('The semi magic square is ...')
        checkSemiM=True
        printbox()
        while not checkOrdinaryMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x):
            box[1][0]+=1
            createMagic()
            if box[1][0]>x-(box[0][0]+box[0][1]):
                checkOrdinaryM=False
                print('Ordinary magic square not found ...')
                break
        if checkOrdinaryMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x):
            checkOrdinaryM=True
            print('The Ordinary Magic square is ...')
            printbox()
    else:
        print('Semi Magic Square not found ...')
    if checkPanMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x):
        print('The Pan Magic Square is ...')
        checkPanM=True
        printbox()
    else:
        print('Pan magic square not found ...')
    if checkCompleteMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x):
        print('The complete Magic Square is ...')
        checkCompleteM=True
        printbox()
    else:
        print('Complete Magic Square Not found ...')
    if checkMPMS(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x):
        print('The Most Perfect Magic Square is ...')
        checkM=True
        printbox()
    else:
        print('Most perfect Magic Square Not found ...')
    if checkSemiM == True and checkPanM == True and checkCompleteM == True and checkOrdinaryM == True and checkM == True:
        print('Your Birthday Satisfies All Magic Square conditions ...\nThe Final Most Perfect Magic Square is :: ')
        printbox()
if createglobals():
    createbox()
    createMagic()
    checkRemaining()

File: magic.js
let box = [
	[0,0,0,0],
	[0,0,0,0],
	[0,0,0,0],
	[0,0,0,0]
]

let form = document.querySelector("form");
let dob = document.getElementById("dob");
let a1 =document.getElementById("a1");
let b1 =document.getElementById("b1");
let c1 =document.getElementById("c1");
let d1 =document.getElementById("d1");
let a2 =document.getElementById("a2");
let b2 =document.getElementById("b2");
let c2 =document.getElementById("c2");
let d2 =document.getElementById("d2");
let a3 =document.getElementById("a3");
let b3 =document.getElementById("b3");
let c3 =document.getElementById("c3");
let d3 =document.getElementById("d3");
let a4 =document.getElementById("a4");
let b4 =document.getElementById("b4");
let c4 =document.getElementById("c4");
let d4 =document.getElementById("d4");

form.addEventListener("submit", function(event) {
    event.preventDefault();
    dob = document.getElementById("dob")

    assignValues()
    dob=""
    createMagic()
    checkRemaining()
    reset()
});
let x=0;
function printbox(){
    document.getElementById("a1").textContent = box[0][0];
    document.getElementById("b1").textContent = box[0][1];
    document.getElementById("c1").textContent = box[0][2];
    document.getElementById("d1").textContent = box[0][3];

    document.getElementById("a2").textContent = box[1][0];
    document.getElementById("b2").textContent = box[1][1];
    document.getElementById("c2").textContent = box[1][2];
    document.getElementById("d2").textContent = box[1][3];

    document.getElementById("a3").textContent = box[2][0];
    document.getElementById("b3").textContent = box[2][1];
    document.getElementById("c3").textContent = box[2][2];
    document.getElementById("d3").textContent = box[2][3];

    document.getElementById("a4").textContent = box[3][0];
    document.getElementById("b4").textContent = box[3][1];
    document.getElementById("c4").textContent = box[3][2];
    document.getElementById("d4").textContent = box[3][3];
}

function assignValues(){
	box[0][0] = parseInt(dob.value[8]+dob.value[9])
	if (dob.value[6]==0){
		box[0][1] = parseInt(dob.value[6])
	}else{
		box[0][1] = parseInt(dob.value[6]+dob.value[7])
	}
	box[0][2] = parseInt(dob.value[0]+dob.value[1])
	if (dob.value[3]==0){
		box[0][3] = parseInt(dob.value[4])
	}else{
		box[0][3] = parseInt(dob.value[3]+dob.value[4])
	}
	return true
	
}
function checkDuplicate(a,n){
	for(let i=0;i<n;i++){
		for(let j=0;j<4;j++){
			if (box[i][j]==a)
				return true
		}
	}
	return false
}

function checkSemiMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x){
    if (a0+b0+c0+d0==x && a1+b1+c1+d1==x && a2+b2+c2+d2==x && a3+b3+c3+d3==x && a0+a1+a2+a3==x && b0+b1+b2+b3==x && c0+c1+c2+c3==x && d0+d1+d2+d3==x)
        return true
    return false
}

function checkOrdinaryMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x){
    if (a0+b1+c2+d3==x && d0+c1+b2+a3==x)
        return true
    return false
}

function checkMPMS(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x){
    if (a0+b0+a1+b1==x && b0+c0+b1+c1==x && c1+c0+d0+d1==x && a2+b2+a1+b1==x && c1+c2+b1+b2==x && c1+c2+d1+d2==x && a2+b2+a3+b3==x && c2+c3+d2+d3==x && b2+b3+c2+c3==x && a0+b0+a3+b3==x && b0+c0+b3+c3==x && c0+d0+c3+d3==x && a0+a1+d0+d1==x && a1+a2+d1+d2==x && a2+a3+d2+d3==x)
        return true
    return false
}

function checkPanMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x){
    if (c0+b1+a2+d3==x && d1+c2+b3+a0==x && b0+c1+d2+a3==x && a1+b2+c3+d0==x)
        return true
    return false
}

function checkCompleteMagic(a0,b0,c0,d0,a1,b1,c1,d1,a2,b2,c2,d2,a3,b3,c3,d3,x){
    if (a0+c0+a2+c2==x && b0+d0+b2+d2==x && a1+c1+a3+c3==x && b1+d1+b3+d3==x)
        return true
    return false
}

function reset(){
    box[1]=[0,0,0,0]
    box[2]=[0,0,0,0]
    box[3]=[0,0,0,0]
}
function createMagic(){
    let l=[],z = false
    x = box[0][0] + box[0][1] + box[0][2] + box[0][3]
    while (box[1][0]<x-(box[0][0]+box[0][1])){
        box[1][1]=x-(box[0][0]+box[0][1]+box[1][0])
        box[1][2]=x-(box[0][2]+box[1][1]+box[0][1])
        box[1][3]=x-(box[0][3]+box[1][2]+box[0][2])
        l.push(...[box[1][0],box[1][1],box[1][2],box[1][3]])
        if ( checkDuplicate(l[0],1) || checkDuplicate(l[1],1) || checkDuplicate(l[2],1) || checkDuplicate(l[3],1) || l.some(i => i < 0) || l.length > new Set(l).size){
            l.length = 0
        }else{  
            l.length = 0
            box[2][0]=0
            while (box[2][0]<x-(box[1][0]+box[1][1])){
                box[2][1]=x-(box[1][0]+box[1][1]+box[2][0])
                box[2][2]=x-(box[1][2]+box[2][1]+box[1][1])
                box[2][3]=x-(box[1][3]+box[2][2]+box[1][2])
                l.push(...[box[2][0],box[2][1],box[2][2],box[2][3]])
                if ( checkDuplicate(l[0],2) || checkDuplicate(l[1],2) || checkDuplicate(l[2],2) || checkDuplicate(l[3],2) || l.some(i => i < 0) || l.length > new Set(l).size){
                    l.length = 0
                }else{
                    l.length = 0
                    box[3][0]=x-(box[2][0]+box[1][0]+box[0][0])
                    box[3][1]=x-(box[2][1]+box[1][1]+box[0][1])
                    box[3][2]=x-(box[2][2]+box[1][2]+box[0][2])
                    box[3][3]=x-(box[2][3]+box[1][3]+box[0][3])
                    l.push(...[box[3][0],box[3][1],box[3][2],box[3][3]])
                    if ( checkSemiMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x) && !(checkDuplicate(box[3][0],3) || checkDuplicate(box[3][1],3) || checkDuplicate(box[3][2],3) || checkDuplicate(box[3][3],3)) && !l.some(i => i < 0) && l.length === new Set(l).size){
                        z=true
                        break	
                    }else{
                    	
                    }
                       	
				}
                box[2][0]+=1
			}
		}
        if (z==true)
            break
        else
            box[1][0]+=1
    }
}
function checkRemaining(){
    let checkSemiM=false;checkOrdinaryM=false;checkM=false;checkPanM=false;checkCompleteM=false
    if (checkSemiMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x)){
        console.log('The semi magic square is ...')
        checkSemiM=true
        while (!checkOrdinaryMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x)){
            box[1][0]+=1
            createMagic()
            if (box[1][0]>x-(box[0][0]+box[0][1])){
                checkOrdinaryM=false
                console.log('Ordinary magic square not found ...')
                break
            }
        }
        if (checkOrdinaryMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x)){
            checkOrdinaryM=true
            console.log('The Ordinary Magic square is ...')
       	}
    }
    else{
        console.log('Semi Magic Square not found ...')
    }
    if (checkPanMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x)){
        console.log('The Pan Magic Square is ...')
        checkPanM=true
    }else{
        console.log('Pan magic square not found ...')
    }
    if (checkCompleteMagic(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x)){
        console.log('The complete Magic Square is ...')
        checkCompleteM=true
        printbox()
    }else{
        console.log('Complete Magic Square Not found ...')
    }if (checkMPMS(box[0][0],box[0][1],box[0][2],box[0][3],box[1][0],box[1][1],box[1][2],box[1][3],box[2][0],box[2][1],box[2][2],box[2][3],box[3][0],box[3][1],box[3][2],box[3][3],x)){
        console.log('The Most Perfect Magic Square is ...')
        checkM=true
    }else{
        console.log('Most perfect Magic Square Not found ...')
    }
    if (checkSemiM == true && checkPanM == true && checkCompleteM == true && checkOrdinaryM == true && checkM == true){
        console.log('Your Birthday Satisfies All Magic Square conditions ...\nThe Final Most Perfect Magic Square is :: ')
        printbox()
    }
}
